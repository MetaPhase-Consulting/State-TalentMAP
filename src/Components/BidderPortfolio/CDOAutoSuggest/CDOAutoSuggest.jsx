import React, { Component } from 'react';
import Dropdown, { DropdownContent } from 'react-simple-dropdown';
import FA from 'react-fontawesome';
import AutoSuggest from '../../AutoSuggest';
import SuggestionChoiceCDOName from '../../AutoSuggest/SuggestionChoiceCDOName';
import { getCurrentUser, filterUsers } from '../temporyHelpers';
import BoxShadow from '../../BoxShadow';
import InteractiveElement from '../../InteractiveElement';

// we'll reference this to return focus after making a selection
const dropdownID = 'cdo-portfolio-dropdown-trigger';

// we'll check for outside clicks and close the dropdown
const dropdownContentID = 'cdo-portfolio-dropdown-content';
const dropdownTriggerID = 'cdo-portfolio-dropdown-trigger';

export const getDisplayProperty = o => `${o.first_name} ${o.last_name}`;

class CDOAutoSuggest extends Component {
  constructor(props) {
    super(props);
    this.onSuggestionSelected = this.onSuggestionSelected.bind(this);
    this.getFilteredUsers = this.getFilteredUsers.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.dropdown = {};
    this.state = {
      suggestions: filterUsers(''),
      selection: getCurrentUser(),
      isActive: false,
    };
  }

  componentWillMount() {
    window.addEventListener('click', this.handleOutsideClick);
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.handleOutsideClick);
  }

  onSuggestionSelected(selection) {
    this.setState({ selection }, () => {
      this.hideDropdown();
      document.getElementById(dropdownID).focus();
    });
  }

  getFilteredUsers(term) {
    this.setState({ suggestions: filterUsers(term) });
  }

  hideDropdown() {
    // Explicitly hide the dropdown using the built-in hide() function from react-simple-dropdown
    if (this.dropdown && this.dropdown.hide) { this.dropdown.hide(); }
    this.setState({ isActive: false });
  }

  showDropdown() {
    // Explicitly show the dropdown using the built-in show() function from react-simple-dropdown
    if (this.dropdown && this.dropdown.show) { this.dropdown.show(); }
    this.setState({ isActive: true });
  }

  toggleDropdown() {
    const { isActive } = this.state;
    if (isActive) {
      this.hideDropdown();
    } else {
      this.showDropdown();
    }
  }

  handleOutsideClick(e) {
    const { isActive } = this.state;
    if (isActive &&
      !document.getElementById(dropdownContentID).contains(e.target)
      && !document.getElementById(dropdownTriggerID).contains(e.target)
    ) {
      this.hideDropdown();
    }
  }

  render() {
    const { isActive, selection, suggestions } = this.state;
    let triggerLabel = `${selection.first_name} ${selection.last_name}`;
    let icon = 'chevron-down';
    let altPrefix = 'Open';
    if (isActive) {
      triggerLabel = 'See client list as';
      icon = 'chevron-up';
      altPrefix = 'Close';
    }
    const alt = `${altPrefix} client list search`;
    return (
      <Dropdown
        className="portfolio-account-dropdown"
        ref={(dropdown) => { this.dropdown = dropdown; }}
        removeElement
        active={isActive}
      >
        <h2>Client Profiles</h2>
        <div className="usa-grid-full dropdown-trigger-container" id={dropdownTriggerID}>
          <span>Client list:</span>
          <InteractiveElement
            onClick={this.toggleDropdown}
            className="account-dropdown--name"
            id={dropdownID}
            alt={alt}
          >
            {triggerLabel}
            <FA name={icon} />
          </InteractiveElement>
        </div>
        <BoxShadow>
          <div className="dropdown-content-outer-container" id={dropdownContentID}>
            <DropdownContent>
              <div className="autosuggest-container">
                <AutoSuggest
                  label="Client list"
                  inputId="cdo-portfolio-autosuggest"
                  debounceMillis={0}
                  suggestions={suggestions}
                  getSuggestions={this.getFilteredUsers}
                  displayProperty={getDisplayProperty}
                  suggestionTemplate={SuggestionChoiceCDOName}
                  onSuggestionSelected={this.onSuggestionSelected}
                  placeholder="Start typing CDO name"
                  autoSuggestProps={{
                    shouldRenderSuggestions: () => true,
                    alwaysRenderSuggestions: true,
                  }}
                  templateProps={{
                    value: selection,
                  }}
                />
              </div>
            </DropdownContent>
          </div>
        </BoxShadow>
      </Dropdown>
    );
  }
}

export default CDOAutoSuggest;
