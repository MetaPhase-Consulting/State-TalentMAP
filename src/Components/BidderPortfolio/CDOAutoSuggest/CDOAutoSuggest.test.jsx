import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CDOAutoSuggest, { getDisplayProperty } from './CDOAutoSuggest';

describe('CDOAutoSuggest', () => {
  it('is defined', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    expect(wrapper).toBeDefined();
  });

  it('sets state on onSuggestionSelected()', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    global.document.getElementById = () => ({ focus: () => {} });
    wrapper.instance().onSuggestionSelected({ id: 1 });
    expect(wrapper.instance().state.isActive).toBe(false);
  });

  it('sets state on getFilteredUsers()', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    wrapper.instance().getFilteredUsers('daniels');
    expect(wrapper.instance().state.suggestions.length).toBe(1);
  });

  it('sets state on hideDropdown()', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    wrapper.instance().hideDropdown();
    expect(wrapper.instance().state.isActive).toBe(false);
  });

  it('sets state on showDropdown()', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    wrapper.instance().showDropdown();
    expect(wrapper.instance().state.isActive).toBe(true);
  });

  it('toggles state on toggleDropdown()', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    wrapper.instance().toggleDropdown();
    expect(wrapper.instance().state.isActive).toBe(true);
    wrapper.instance().toggleDropdown();
    expect(wrapper.instance().state.isActive).toBe(false);
    wrapper.instance().toggleDropdown();
    expect(wrapper.instance().state.isActive).toBe(true);
  });

  it('toggles state on toggleDropdown()', () => {
    const output = 'John Smith';
    expect(getDisplayProperty({ first_name: 'John', last_name: 'Smith' })).toBe(output);
  });

  it('calls this.hideDropdown on this.handleOutsideClick when !!isActive and element does not contain target', () => {
    window.document.getElementById = () => ({
      contains: () => false,
    });
    const wrapper = shallow(<CDOAutoSuggest />);
    wrapper.instance().setState({ isActive: true });
    wrapper.update();
    wrapper.instance().handleOutsideClick({ target: 1 });
    expect(wrapper.instance().state.isActive).toBe(false);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('matches snapshot after toggling', () => {
    const wrapper = shallow(<CDOAutoSuggest />);
    wrapper.find('InteractiveElement').simulate('click');
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
