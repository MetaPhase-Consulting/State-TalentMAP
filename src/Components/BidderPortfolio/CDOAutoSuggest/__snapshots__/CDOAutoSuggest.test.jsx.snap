// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDOAutoSuggest matches snapshot 1`] = `
<Dropdown
  active={false}
  className="portfolio-account-dropdown"
  removeElement={true}
>
  <h2>
    Client Profiles
  </h2>
  <div
    className="usa-grid-full dropdown-trigger-container"
    id="cdo-portfolio-dropdown-trigger"
  >
    <span>
      Client list:
    </span>
    <InteractiveElement
      alt="Open client list search"
      className="account-dropdown--name"
      id="cdo-portfolio-dropdown-trigger"
      onClick={[Function]}
      type="div"
    >
      Mary Brown
      <FontAwesome
        name="chevron-down"
      />
    </InteractiveElement>
  </div>
  <BoxShadow
    blurRadius={10}
    color="rgba(0,0,0,.15)"
    inset={false}
    is="div"
    offsetX={3}
    offsetY={2}
    spreadRadius={1}
    style={Object {}}
  >
    <div
      className="dropdown-content-outer-container"
      id="cdo-portfolio-dropdown-content"
    >
      <DropdownContent
        className=""
      >
        <div
          className="autosuggest-container"
        >
          <AutoSuggest
            autoSuggestProps={
              Object {
                "alwaysRenderSuggestions": true,
                "shouldRenderSuggestions": [Function],
              }
            }
            customInputProps={Object {}}
            debounceMillis={0}
            displayProperty={[Function]}
            getSuggestions={[Function]}
            inputId="cdo-portfolio-autosuggest"
            label="Client list"
            labelSrOnly={true}
            onSuggestionSelected={[Function]}
            onSuggestionsClearRequested={[Function]}
            placeholder="Start typing CDO name"
            queryProperty=""
            shouldClearOnSelect={false}
            suggestionTemplate={[Function]}
            suggestions={
              Array [
                Object {
                  "first_name": "Mary",
                  "id": 4,
                  "last_name": "Brown",
                },
                Object {
                  "first_name": "John",
                  "id": 3,
                  "last_name": "Daniels",
                },
              ]
            }
            templateProps={
              Object {
                "value": Object {
                  "first_name": "Mary",
                  "id": 4,
                  "last_name": "Brown",
                },
              }
            }
          />
        </div>
      </DropdownContent>
    </div>
  </BoxShadow>
</Dropdown>
`;

exports[`CDOAutoSuggest matches snapshot after toggling 1`] = `
<Dropdown
  active={true}
  className="portfolio-account-dropdown"
  removeElement={true}
>
  <h2>
    Client Profiles
  </h2>
  <div
    className="usa-grid-full dropdown-trigger-container"
    id="cdo-portfolio-dropdown-trigger"
  >
    <span>
      Client list:
    </span>
    <InteractiveElement
      alt="Close client list search"
      className="account-dropdown--name"
      id="cdo-portfolio-dropdown-trigger"
      onClick={[Function]}
      type="div"
    >
      See client list as
      <FontAwesome
        name="chevron-up"
      />
    </InteractiveElement>
  </div>
  <BoxShadow
    blurRadius={10}
    color="rgba(0,0,0,.15)"
    inset={false}
    is="div"
    offsetX={3}
    offsetY={2}
    spreadRadius={1}
    style={Object {}}
  >
    <div
      className="dropdown-content-outer-container"
      id="cdo-portfolio-dropdown-content"
    >
      <DropdownContent
        className=""
      >
        <div
          className="autosuggest-container"
        >
          <AutoSuggest
            autoSuggestProps={
              Object {
                "alwaysRenderSuggestions": true,
                "shouldRenderSuggestions": [Function],
              }
            }
            customInputProps={Object {}}
            debounceMillis={0}
            displayProperty={[Function]}
            getSuggestions={[Function]}
            inputId="cdo-portfolio-autosuggest"
            label="Client list"
            labelSrOnly={true}
            onSuggestionSelected={[Function]}
            onSuggestionsClearRequested={[Function]}
            placeholder="Start typing CDO name"
            queryProperty=""
            shouldClearOnSelect={false}
            suggestionTemplate={[Function]}
            suggestions={
              Array [
                Object {
                  "first_name": "Mary",
                  "id": 4,
                  "last_name": "Brown",
                },
                Object {
                  "first_name": "John",
                  "id": 3,
                  "last_name": "Daniels",
                },
              ]
            }
            templateProps={
              Object {
                "value": Object {
                  "first_name": "Mary",
                  "id": 4,
                  "last_name": "Brown",
                },
              }
            }
          />
        </div>
      </DropdownContent>
    </div>
  </BoxShadow>
</Dropdown>
`;
